### PR TITLE
Introduce anonymous product analytics

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | This PR introduces anonymous analytics to know how many times the module is used to upgrade / restore and how.
| Type?             | new feature
| BC breaks?        | Yes, some files are being moved, i.e Tasks
| Deprecations?     | Nope
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/discussions/36432
| Sponsor company   | @PrestaShopCorp
| How to test?      | Triggering updates and restores will be counted, as well as their ending (Successful or not).

### CODE REVIEW

**Review by matks — APPROVED** [review]

File: classes/State.php [hunk 1]
```diff
@@ -322,6 +338,18 @@ public function setRestoreDbFilenames(array $restoreDbFilenames): State
         return $this;
     }
 
+    public function getRestoreVersion(): ?string
+    {
+        $matches = [];
+        preg_match(
+            '/^V(?<version>[1-9\.]+)_/',
```
[the remark this answers is not in the document]
Comment (Quetzacoalt91) [reply classes/State.php:345]: Not sure I got your message properly. What do you mean by excluding when I try to pickup the version from a string?

[the remark this answers is not in the document]
Comment (matks) [reply classes/State.php:345]: If I call getRestoreVersion(), you check `$this->getRestoreName()` and verify if it matches `V(?<version>[1-9\.]+)`

If it matches, you return it. If it does not you return null. So I said you "exclude" patterns which do not match.

An implementation that does not do exclusion would be simply returning `$this->getRestoreName()

I suggest explaining the reason behind the above logic

[the remark this answers is not in the document]
Comment (Quetzacoalt91) [reply classes/State.php:348]: Here you go

File: classes/Analytics.php [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    public const URL_TRACKING_ENV_NAME = 'PS_URL_TRACKING';
+
+    /**
+     * @var string
+     */
+    private $anonymousId;
+
+    /**
+     * @var array
+     */
```
Comment (M0rgan01) [inline classes/Analytics.php:51]: This will not work with the new level of PHPstant :cry:

Comment (Quetzacoalt91) [reply classes/Analytics.php:51]: Fortunately PHPStan's PR should be merged first, and I'll get all the new feedback from it.

File: classes/Task/Runner/ChainedTasks.php [hunk 3]
```diff
@@ -55,13 +57,19 @@ public function run(): int
...
+                $controller->setErrorFlag();
+                throw $t;
+            }
 
             $result = $controller->getResponse();
             $requireRestart = $this->checkIfRestartRequested($result);
             $this->error = $result->getError();
             $this->stepDone = $result->getStepDone();
-            $this->step = $result->getNext();
+            $this->step = $this->next = $result->getNext();
```
Comment (M0rgan01) [inline classes/Task/Runner/ChainedTasks.php:71]:
```suggestion
            $this->step = $result->getNext();
```

**Review by M0rgan01 — APPROVED** [review]

**Comment by Quetzacoalt91** [comment]
It appeared during the tests that failing upgrade are not reported in case of a PHP timeout. There is room for improvement.

**Comment by matks** [comment]
@Quetzacoalt91 Maybe it's the moment where `register_shutdown_function()` shines

**Comment by Quetzacoalt91** [comment]
Indeed, we already have it to send responses gracefully, so it may be completed with this feature.

**Pull request merged**